crates/fuel-core/src/database/balances.rs

```diff
@@ -48,37 +59,91 @@ impl StorageMutate<ContractsAssets> for Database {
         key: &<ContractsAssets as Mappable>::Key,
         value: &<ContractsAssets as Mappable>::Value,
     ) -> Result<Option<<ContractsAssets as Mappable>::OwnedValue>, Self::Error> {
-        Database::insert(self, key.as_ref(), Column::ContractsAssets, value)
-            .map_err(Into::into)
+        let prev = Database::insert(self, key.as_ref(), Column::ContractsAssets, value)
+            .map_err(Into::into);
+
+        // Get latest metadata entry for this contract id
+        let prev_metadata = self
+            .storage::<ContractsAssetsMerkleMetadata>()
+            .get(key.contract_id())?
+            .unwrap_or_default();
+
+        let root = prev_metadata.root;
+        let mut tree: MerkleTree<ContractsAssetsMerkleData, _> = {
+            let storage = self.borrow_mut();
+            if root == [0; 32] {
```

Comment:
It sounds like you're suggesting that we update the functionality of `MerkleTree::load` to something like:
- If the root is the empty root, return a new tree (without loading anything)
- Otherwise, load the existing tree from the root

That could work! I will create a PR for that in fuel-vm to see how that looks. 

However, I would say that this is not an internal detail because this is mandated by the spec: https://github.com/celestiaorg/celestia-specs/blob/master/src/specs/data_structures.md#sparse-merkle-tree. All spec-compliant Merkle trees have this quality.

Reply:
I meant from the `fuel-core` perspective it is internal details of the `fuel-merkle`=)

Reply:
PR up: https://github.com/FuelLabs/fuel-vm/pull/379. 
This will simplify the above code to something like:
```
let root = prev_metadata.root;
let storage = self.borrow_mut();
let mut tree: MerkleTree<ContractsAssetsMerkleData, _> = MerkleTree::load(storage, root)...
```